### PR TITLE
core/types: use rlp to encode/decode istanbul extra-data

### PR DESCRIPTION
**Pull request by bailantaotao**

### CODE REVIEW

File: consensus/istanbul/backend/engine.go [hunk 1]
```diff
@@ -620,23 +632,23 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
 func sigHash(header *types.Header) (hash common.Hash) {
 	hasher := sha3.NewKeccak256()
 
-	index := types.ExtractToIstanbulIndex(header)
+	newHeader := types.IstanbulExtraDataFilter(header)
 	rlp.Encode(hasher, []interface{}{
-		header.ParentHash,
```
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: use newHeader in rlp.Encode

File: consensus/istanbul/backend/engine.go [hunk 2]
```diff
@@ -455,8 +462,10 @@ func (sb *simpleBackend) updateBlock(parent *types.Header, block *types.Block) (
 		return nil, err
 	}
 
-	index := types.ExtractToIstanbulIndex(header)
-	copy(header.Extra[index.Seal:index.Seal+types.IstanbulExtraSeal], sighash)
+	err = writeIstanbulSeal(header, sighash)
```
Comment (yutelin) [inline consensus/istanbul/backend/engine.go]: rename `sighash` variable to `seal`

File: core/types/istanbul.go [hunk 3]
```diff
@@ -17,199 +17,87 @@
...
-// may be insufficient, so we will append 0x00 to avoid panic in getting array.
-//
-// Extra-data format like below.
-// ┌─────────────────┬──────────────────────┬────────────────────────────────┬───────────────┬─────────────────────┬──────────────────────────────┐
-// │ vanity(32 bytes)│ validator#N (1 bytes)│ validator address(N * 20 bytes)│ seal(65 bytes)│ committed#N (1 byte)│ committed seals(N * 65 bytes)│
-// └─────────────────┴──────────────────────┴────────────────────────────────┴───────────────┴─────────────────────┴──────────────────────────────┘
-func ensureValidIstanbulExtra(h *Header) *Header {
+// IstanbulExtraDataFilter is used to filter out the useless information(like seal, committed seals) to
+// ensure it conforms to istanbul hash rules. It returns nil if extra-data field of header cannot
+// decode/encode by rlp.
```
Comment (markya0616) [inline core/types/istanbul.go]: IstanbulFilteredHeader returns a filtered header which some information (like seal, committed seals) are clean to fulfill the Istanbul hash rules. It returns nil if the extra data of header cannot be decoded/encoded by rlp.

File: core/types/istanbul.go [hunk 4]
```diff
@@ -17,199 +17,87 @@
...
+func IstanbulExtraDataFilter(h *Header) *Header {
```
Comment (markya0616) [inline core/types/istanbul.go]: Suggest to rename to IstanbulFilteredHeader

File: core/types/istanbul.go [hunk 5]
```diff
@@ -17,199 +17,87 @@
...
-func PrepareIstanbulExtra(header *Header, vals []common.Address) []byte {
-	var buf bytes.Buffer
-
-	// 1. compensate the lack bytes if header.Extra is not enough IstanbulExtraVanity bytes.
-	if len(header.Extra) < IstanbulExtraVanity {
-		header.Extra = append(header.Extra, bytes.Repeat([]byte{0x00}, IstanbulExtraVanity-len(header.Extra))...)
+// ExtractToIstanbul extracts all values of the IstanbulExtra from the header. It returns an
+// error if the passed length of header extra-data is less than 32 or the header extra can not
+// decode.
+func ExtractToIstanbul(h *Header) (*IstanbulExtra, error) {
```
Comment (markya0616) [inline core/types/istanbul.go]: Can we rename to IstanbulExtra? 
// IstanbulExtra returns a IstanbulExtra from the given header.

File: core/types/istanbul.go [hunk 6]
```diff
@@ -17,199 +17,87 @@
...
-	IstanbulExtraVanity        = int(params.MaximumExtraDataSize) // Fixed number of extra-data bytes reserved for validator vanity
-	IstanbulExtraValidatorSize = 1                                // Fixed number of extra-data bytes reserved for validator size
-	IstanbulExtraSeal          = 65                               // Fixed number of extra-data bytes reserved for validator seal
-	IstanbulExtraCommittedSize = 1                                // Fixed number of extra-data bytes reserved for committed size
+	IstanbulExtraVanity = 32 // Fixed number of extra-data bytes reserved for validator vanity
+	IstanbulExtraSeal   = 65 // Fixed number of extra-data bytes reserved for validator seal
 
-	ErrInvalidIstanbulHeaderExtra   = fmt.Errorf("Invalid istanbul header extra-data")
-	ErrInvalidIstanbulCommittedSeal = fmt.Errorf("Invalid istanbul committed seal")
+	ErrInvalidIstanbulHeaderExtra = fmt.Errorf("Invalid istanbul header extra-data")
```
Comment (markya0616) [inline core/types/istanbul.go]: use errors.New and add comments

File: core/types/block.go [hunk 7]
```diff
@@ -100,7 +100,11 @@ type headerMarshaling struct {
 // RLP encoding.
 func (h *Header) Hash() common.Hash {
 	if h.MixDigest == IstanbulDigest {
-		return rlpHash(IstanbulHashFilter(h))
+		if istanbulHeader := IstanbulExtraDataFilter(h); istanbulHeader != nil {
+			return rlpHash(istanbulHeader)
+		} else {
+			return common.Hash{}
```
Comment (markya0616) [inline core/types/block.go]: can we return the original one? `rlpHash(h)`

File: consensus/istanbul/backend/engine.go [hunk 8]
```diff
@@ -620,32 +632,89 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
...
+	istExtra, err := types.ExtractToIstanbul(header)
+	if err != nil {
+		return common.Address{}, err
+	}
+	return istanbul.GetSignatureAddress(sigHash(header).Bytes(), istExtra.Seal)
+}
+
+// prepareIstanbulExtra returns a istanbul extra-data that use the given header, parent to
+// ensure it conforms to Istanbul rules, but seal, and committed seals are not included.
+func prepareIstanbulExtra(header *types.Header, vals []common.Address) ([]byte, error) {
```
Comment (markya0616) [inline consensus/istanbul/backend/engine.go]: We've already in istanbul package, so we may rename to prepareExtra.

File: consensus/istanbul/backend/engine.go [hunk 9]
```diff
@@ -620,32 +632,89 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
...
-	return istanbul.GetSignatureAddress(sigHash(header).Bytes(), signature)
+	istExtra, err := types.ExtractToIstanbul(header)
+	if err != nil {
+		return common.Address{}, err
+	}
+	return istanbul.GetSignatureAddress(sigHash(header).Bytes(), istExtra.Seal)
+}
+
+// prepareIstanbulExtra returns a istanbul extra-data that use the given header, parent to
+// ensure it conforms to Istanbul rules, but seal, and committed seals are not included.
```
Comment (markya0616) [inline consensus/istanbul/backend/engine.go]: prepareExtra returns a extra-data of the given header and validators

File: consensus/istanbul/backend/engine.go [hunk 10]
```diff
@@ -620,32 +632,89 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
...
+	if err != nil {
+		return err
+	}
+
+	h.Extra = append(h.Extra[:types.IstanbulExtraVanity], payload...)
+	return nil
+}
+
+// writeIstanbulCommittedSeals writes the extra-data field of a block header with given committed seals.
+func writeIstanbulCommittedSeals(h *types.Header, committedSeals []byte) error {
```
Comment (markya0616) [inline consensus/istanbul/backend/engine.go]: rename to writeCommittedSeals

File: core/types/istanbul.go (hunk 3 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: ok

File: consensus/istanbul/backend/engine.go (hunk 2 above)
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: ok

File: core/types/istanbul.go (hunk 4 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: OK

File: core/types/istanbul.go (hunk 5 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: ok

File: core/types/istanbul.go (hunk 6 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: OK

File: consensus/istanbul/backend/engine.go (hunk 8 above)
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: OK

File: consensus/istanbul/backend/engine.go [hunk 11]
```diff
@@ -620,32 +632,89 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
...
+	payload, err := rlp.EncodeToBytes(&ist)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(buf.Bytes(), payload...), nil
+}
+
+// writeIstanbulSeal writes the extra-data field of a block header with given seal.
+func writeIstanbulSeal(h *types.Header, seal []byte) error {
```
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: OK

File: consensus/istanbul/backend/engine.go (hunk 9 above)
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: OK

File: consensus/istanbul/backend/engine.go (hunk 10 above)
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: OK

File: core/types/block.go (hunk 7 above)
Comment (bailantaotao) [inline core/types/block.go]: OK

File: consensus/istanbul/backend/engine.go [hunk 12]
```diff
@@ -253,7 +253,10 @@ func (sb *simpleBackend) verifyCommittedSeals(chain consensus.ChainReader, heade
 		return err
 	}
 
-	extra := types.ExtractToIstanbul(header)
+	extra, err := types.ExtractToIstanbul(header)
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: How about renaming to `ExtractIstanbulExtra`?

File: consensus/istanbul/backend/engine.go [hunk 13]
```diff
@@ -562,7 +571,10 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
 			if err := sb.VerifyHeader(chain, genesis, false); err != nil {
 				return nil, err
 			}
-			istanbul := types.ExtractToIstanbul(genesis)
+			istanbul, err := types.ExtractToIstanbul(genesis)
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: `istanbul` -> `istanbulExtra`

File: consensus/istanbul/backend/engine.go [hunk 14]
```diff
@@ -620,32 +632,89 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
...
 	return hash
 }
 
 // ecrecover extracts the Ethereum account address from a signed header.
 func ecrecover(header *types.Header) (common.Address, error) {
 	// Retrieve the signature from the header extra-data
-	index := types.ExtractToIstanbulIndex(header)
-	signature := header.Extra[index.Seal : index.Seal+types.IstanbulExtraSeal]
-	return istanbul.GetSignatureAddress(sigHash(header).Bytes(), signature)
+	istExtra, err := types.ExtractToIstanbul(header)
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: Variable name should be consistent with previous one.
`istExtra` -> `istanbulExtra`

File: consensus/istanbul/backend/engine.go [hunk 15]
```diff
@@ -620,32 +632,89 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
...
+	return append(buf.Bytes(), payload...), nil
+}
+
+// writeIstanbulSeal writes the extra-data field of a block header with given seal.
+func writeIstanbulSeal(h *types.Header, seal []byte) error {
+	if len(seal)%types.IstanbulExtraSeal != 0 {
+		return errInvalidSignature
+	}
+
+	ist, err := types.ExtractToIstanbul(h)
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: `ist` -> `istanbulExtra`

File: consensus/istanbul/backend/engine.go [hunk 16]
```diff
@@ -620,32 +632,89 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
...
+	return nil
+}
+
+// writeIstanbulCommittedSeals writes the extra-data field of a block header with given committed seals.
+func writeIstanbulCommittedSeals(h *types.Header, committedSeals []byte) error {
+	if len(committedSeals)%types.IstanbulExtraSeal != 0 {
+		return errInvalidCommittedSeals
+	}
+
+	ist, err := types.ExtractToIstanbul(h)
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: `ist` -> `istanbulExtra`

File: consensus/istanbul/backend/engine.go (hunk 10 above)
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: How about using `committedSeals [][]byte` here?
In `core`, seals were made up by appending many byte arrays. But in here, we split this byte array and copy the contents to `CommittedSeal`.

File: consensus/istanbul/backend/engine.go [hunk 17]
```diff
@@ -620,32 +632,89 @@ func (sb *simpleBackend) snapshot(chain consensus.ChainReader, number uint64, ha
...
+	if err != nil {
+		return err
+	}
+
+	ist.CommittedSeal = make([][]byte, len(committedSeals)/types.IstanbulExtraSeal)
```
Comment (alanchchen) [inline consensus/istanbul/backend/engine.go]: I think it should be `CommittedSeals`.
```
type IstanbulExtra struct {
	Vanity         []byte
	Validators     []common.Address
	Seal           []byte
	CommittedSeals [][]byte
}
```

File: consensus/istanbul/backend/engine.go (hunk 12 above)
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: OK

File: consensus/istanbul/backend/engine.go (hunk 13 above)
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: OK

File: consensus/istanbul/backend/engine.go (hunk 15 above)
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: OK

File: consensus/istanbul/backend/engine.go (hunk 14 above)
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: OK

File: consensus/istanbul/backend/engine.go (hunk 16 above)
Comment (bailantaotao) [inline consensus/istanbul/backend/engine.go]: OK

**Pull request merged**